### PR TITLE
disable custom metrics

### PR DESCRIPTION
**Pull request by pfrcks**
This pull request primarily introduces the ability to enable custom metrics in the system. It also makes adjustments to the conditions under which resource optimization is performed and modifies several configuration files. 

The key changes are as follows:

Environment Variable Introduction:

* [`build/common/installer/scripts/fluent-bit-conf-customizer.rb`](diffhunk://#diff-9ddf023aaa9b5992b30250db0faebd74e2ee801c09a1d1cc46f6ec7781bf9a16R134): Introduced a new environment variable `ENABLE_CUSTOM_METRICS` to enable or disable custom metrics.

Condition Modification:

* [`build/common/installer/scripts/fluent-bit-conf-customizer.rb`](diffhunk://#diff-9ddf023aaa9b5992b30250db0faebd74e2ee801c09a1d1cc46f6ec7781bf9a16L197-R201): Modified the condition for resource optimization. Now, resource optimization will be performed if the system is Linux and custom metrics are not enabled, or if the system is Windows and Fluent Bit is not disabled.

Disabling Resource Optimization:

* [`build/common/installer/scripts/tomlparser-agent-config.rb`](diffhunk://#diff-fa01e80dbcfab688d985fab24ffe79cc94d5246e1cc10a55cf477c0146dabd43R387-R396): If custom metrics are enabled, resource optimization is disabled.

Custom Metrics Enabling:

* [`build/common/installer/scripts/tomlparser-common-agent-config.rb`](diffhunk://#diff-8febaeaf011c6173c3f18a49405a41e1a4bf6896d94872c35ff258d50b3d0141R85-R92): Added the ability to enable custom metrics through the parsed configuration.

Configuration File Changes:

* `build/linux/installer/conf/container-cm.conf`, `build/linux/installer/conf/kube-cm.conf`: Added new configurations for custom metrics. [[1]](diffhunk://#diff-679d087b044d84191ea26bcd8a521022facf28b874f7597b751be003597fa90bR1-R141) [[2]](diffhunk://#diff-b67ef7026f9725c7d3093b529001fc6545110cd20d3f4ec2fa173631e76560fbR1-R358)
* `build/linux/installer/conf/container.conf`, `build/linux/installer/conf/kube.conf`: Removed configurations related to custom metrics. [[1]](diffhunk://#diff-e500149c08ad346f4135f753d5d0ae4ec6eb9dd2c2d84ee3db11f768b9694716L11-L19) [[2]](diffhunk://#diff-e500149c08ad346f4135f753d5d0ae4ec6eb9dd2c2d84ee3db11f768b9694716L36-L47) [[3]](diffhunk://#diff-e500149c08ad346f4135f753d5d0ae4ec6eb9dd2c2d84ee3db11f768b9694716L100-L116) [[4]](diffhunk://#diff-ecda4597169f4e3a53863d5aaddcfef190cc4f914a6de42dfd67539e617e2396L62-L69) [[5]](diffhunk://#diff-ecda4597169f4e3a53863d5aaddcfef190cc4f914a6de42dfd67539e617e2396L166-L170) [[6]](diffhunk://#diff-ecda4597169f4e3a53863d5aaddcfef190cc4f914a6de42dfd67539e617e2396L197-L214)

### CODE REVIEW

**Comment by pfrcks** [comment]
@ganga1980 Do I need the changes in charts/azuremonitor-containers in charts/azuremonitor-containers-geneva as well?

File: build/windows/installer/livenessprobe/livenessprobe.cpp [hunk 1]
```diff
@@ -113,13 +113,20 @@ int _tmain(int argc, wchar_t *argv[])
         wprintf_s(L"ERROR:Process:%s is not running\n", argv[1]);
         return NO_FLUENT_BIT_PROCESS;
     }
+    const DWORD bufferSize = 256;
+    wchar_t enableCustomMetricsValue[bufferSize];
```
Comment (ganga1980) [inline build/windows/installer/livenessprobe/livenessprobe.cpp:117]: Do we need 256 bytes for value of environment variables which can true or false or "", right? This is static allocation and essentially, we are reserving 512 bytes here for both environment variables.

File: build/windows/installer/livenessprobe/livenessprobe.cpp [hunk 2]
```diff
@@ -113,13 +113,20 @@ int _tmain(int argc, wchar_t *argv[])
...
+    wchar_t msiModeValue[bufferSize];
+    GetEnvironmentVariable(L"ENABLE_CUSTOM_METRICS", enableCustomMetricsValue, bufferSize);
+    GetEnvironmentVariable(L"USING_AAD_MSI_AUTH", msiModeValue, bufferSize);
 
-    DWORD dwStatus = GetServiceStatus(argv[2]);
-
-    if (dwStatus != SERVICE_RUNNING)
+    if (_wcsicmp(enableCustomMetricsValue, L"true") == 0 || _wcsicmp(msiModeValue, L"true") != 0)
```
Comment (ganga1980) [inline build/windows/installer/livenessprobe/livenessprobe.cpp:122]: please verify this change works.

File: build/linux/installer/conf/kube-cm.conf [hunk 3]
```diff
@@ -0,0 +1,358 @@
+   #fluent forward plugin
+   <system>
+    workers "#{ENV['NUM_OF_FLUENTD_WORKERS']}"
+    root_dir /var/opt/microsoft/docker-cimprov/state
+   </system>
+
+   #@include windows_rs_containerinventory.conf
```
Comment (jatakiajanvi12) [inline build/linux/installer/conf/kube-cm.conf:7]: Why is this in the custom metrics kube file? It should be in the previous file itself

File: kubernetes/windows/main.ps1 [hunk 4]
```diff
@@ -291,39 +292,40 @@ function Set-EnvironmentVariables {
...
-            $proxy = [string]$proxy.Trim();
+    }
+    else {
+        $proxy = ""
+        if (Test-Path /etc/ama-logs-secret/PROXY) {
+            # TODO: Change to ama-logs-secret before merging
+            $proxy = Get-Content /etc/ama-logs-secret/PROXY
+            Write-Host "Validating the proxy configuration since proxy configuration provided"
+            # valide the proxy endpoint configuration
             if (![string]::IsNullOrEmpty($proxy)) {
```
Comment (jatakiajanvi12) [inline kubernetes/windows/main.ps1]: do we need any of these changes here?

File: kubernetes/windows/main.ps1 [hunk 5]
```diff
@@ -852,24 +867,24 @@ function Bootstrap-CACertificates {
 }
 
 function IsGenevaMode() {
-    $isGenevaLogsIntegration=$false
-    $isGenevaLogsMultitenancy=$false
+    $isGenevaLogsIntegration = $false
+    $isGenevaLogsMultitenancy = $false
     $genevaLogsIntegration = [System.Environment]::GetEnvironmentVariable("GENEVA_LOGS_INTEGRATION")
     $genevaLogsMultitenancy = [System.Environment]::GetEnvironmentVariable("GENEVA_LOGS_MULTI_TENANCY")
     $genevaLogsInfraNameSpaces = [System.Environment]::GetEnvironmentVariable("GENEVA_LOGS_INFRA_NAMESPACES")
```
Comment (jatakiajanvi12) [inline kubernetes/windows/main.ps1]: Can we remove all these unnecessary changes? It is becoming confusing to understand what exactly are the changes needed for this feature?

**Review by jatakiajanvi12 — CHANGES_REQUESTED** [review]

File: build/linux/installer/conf/kube-cm.conf (hunk 3 above)
Comment (pfrcks) [reply build/linux/installer/conf/kube-cm.conf:7]: this file's config is what we have now. the file kube.conf is the one without custom metrics. 

Are you saying that this should be present in both files?

Comment (pfrcks) [reply build/linux/installer/conf/kube-cm.conf:7]: I just checked and this line is present in both files. @jatakiajanvi12

**Comment by ganga1980** [comment]
> @ganga1980 Do I need the changes in charts/azuremonitor-containers in charts/azuremonitor-containers-geneva as well?

no, we dont need.

File: build/common/installer/scripts/tomlparser-common-agent-config.rb [hunk 6]
```diff
@@ -11,6 +11,7 @@
 @disableTelemetry = false
 @logEnableKubernetesMetadataCacheTTLSeconds = 60
 @enableHighLogScaleMode = false
+@enableCustomMetrics = false
```
Comment (ganga1980) [inline build/common/installer/scripts/tomlparser-common-agent-config.rb:14]: Change looks good. Few questions
1. Did we test the custom metrics enablement and disablement works? 
2. how much perf improvements we are seeing when custom metrics disabled?
3. what's plan on deprecate the custom metrics for customers who will be enabling through config map settings?

Comment (pfrcks) [reply build/common/installer/scripts/tomlparser-common-agent-config.rb:14]: 1. Tested it for legacy its working. MSI pending
2. Pending
3. After Aug, we will make configmap enablement a no-op

**Comment by pfrcks** [comment]
1. Tested MSI and Legacy Mode
2. Tested enablement of custom metrics via cm
3. Tested updated livenessprobe functionality of windows by manually stopping fluentd in normal scenarios
4. Tested fluentd presence on windows for legacy mode and absence for msi mode

**All comparison in order -> image with disabled custom metrics followed by image 3.1.20**

MSI Linux DS MEM
msi-linux-ds-mem

MSI Linux DS CPU
msi-linux-ds-cpu

MSI Linux RS MEM
msi-linux-rs-mem

MSI Linux RS CPU
msi-linux-rs-cpu


MSI Windows DS WorkingSet Bytes
msi-windows-ds-mem-workingset


MSI Windows DS CPU
msi-windows-ds-cpu






Legacy DS MEM
legacy-linux-ds-mem

Legacy DS CPU
legacy-linux-ds-cpu

Legacy RS MEM
legacy-linux-rs-mem

Legacy RS CPU
legacy-linux-rs-cpu

Legacy Windows WorkingSet
legacy-windows-ds-mem-workingset

Legacy Windows CPU
legacy-windows-ds-cpu





Major improvement in Windows platform for both Legacy and MSI (30 MB improvement in Legacy vs 125 MB improvement in MSI)
Minor improvement in RS for both Legacy and MSI
Similar values in DS for both Legacy and MSI

File: build/common/installer/scripts/tomlparser-common-agent-config.rb (hunk 6 above)
Comment (pfrcks) [reply build/common/installer/scripts/tomlparser-common-agent-config.rb:14]: Added findings below

File: kubernetes/linux/main.sh [hunk 7]
```diff
@@ -1073,19 +1073,32 @@ if [ -e "/opt/dcr_env_var" ]; then
       setGlobalEnvVar LOGS_AND_EVENTS_ONLY "${LOGS_AND_EVENTS_ONLY}"
 fi
 
-setGlobalEnvVar AZMON_RESOURCE_OPTIMIZATION_ENABLED "${AZMON_RESOURCE_OPTIMIZATION_ENABLED}"
+setGlobalEnvVar ENABLE_CUSTOM_METRICS "${ENABLE_CUSTOM_METRICS}"
+if [ "${ENABLE_CUSTOM_METRICS}" == "true" ]; then
```
Comment (jatakiajanvi12) [inline kubernetes/linux/main.sh:1078]: Does this mean if someone has custom metrics on, they will not get the other benefits as well?

File: kubernetes/linux/main.sh [hunk 8]
```diff
@@ -1073,19 +1073,32 @@ if [ -e "/opt/dcr_env_var" ]; then
...
+                              mv /etc/fluent/container-cm.conf /etc/fluent/container.conf
+                        fi
                         fluentd -c /etc/fluent/container.conf -o /var/opt/microsoft/docker-cimprov/log/fluentd.log --log-rotate-age 5 --log-rotate-size 20971520 &
                   else
                         echo "Skipping fluentd since LOGS_AND_EVENTS_ONLY is set to true"
                   fi
             else
                   echo "*** starting fluentd v1 in replicaset"
+                  if [ "${ENABLE_CUSTOM_METRICS}" == "true" ]; then
+                        mv /etc/fluent/kube-cm.conf /etc/fluent/kube.conf
```
Comment (jatakiajanvi12) [inline kubernetes/linux/main.sh:1101]: This move will undo whatever updates we are making for windows_rs changes for perf and container inventory above

Comment (jatakiajanvi12) [inline kubernetes/linux/main.sh:1101]: I think we should test all the different windows cases as well with this PR to make sure nothing is breaking: 
- Legacy and MSI
- MSI with and without AZMON_WINDOWS_FLUENT_BIT_DISABLED
- Geneva (legacy and MSI)

**Review by jatakiajanvi12 — CHANGES_REQUESTED** [review]

File: kubernetes/linux/main.sh (hunk 7 above)
Comment (pfrcks) [reply kubernetes/linux/main.sh:1078]: yes, because resource optimization assumes that custom metrics is not needed. so if cx has enabled custom metrics we cannot enable resource optimization

File: kubernetes/linux/main.sh (hunk 8 above)
Comment (pfrcks) [reply kubernetes/linux/main.sh:1101]: updatd the code to replace in both files

File: build/common/installer/scripts/tomlparser-common-agent-config.rb [hunk 9]
```diff
@@ -11,6 +11,7 @@
 @disableTelemetry = false
 @logEnableKubernetesMetadataCacheTTLSeconds = 60
```
Comment (ganga1980) [inline build/common/installer/scripts/tomlparser-common-agent-config.rb:12]: can you please regenrate PR description since the approach is changed?

File: build/common/installer/scripts/tomlparser-common-agent-config.rb [hunk 10]
```diff
@@ -81,6 +82,14 @@ def populateSettingValuesFromConfigMap(parsedConfig)
       end
     end
```
Comment (ganga1980) [inline build/common/installer/scripts/tomlparser-common-agent-config.rb:84]: as we discussed, read the ENABLE_CUSTOM_METRICS  envvar and if its disabled, return false in this  function https://github.com/microsoft/Docker-Provider/blob/ci_prod/source/plugins/ruby/CustomMetricsUtils.rb#L10 as this being used in input plugins to generate MDM records.

File: source/plugins/ruby/in_kube_podinventory.rb [hunk 11]
```diff
@@ -2,6 +2,7 @@
 # frozen_string_literal: true
 
 require "fluent/plugin/input"
+require_relative "oms_common"
```
Comment (ganga1980) [inline source/plugins/ruby/in_kube_podinventory.rb:5]: why do we need this dependency?

Comment (pfrcks) [reply source/plugins/ruby/in_kube_podinventory.rb:5]: `module Fluent::Plugin
  class Kube_PodInventory_Input < Input
    Fluent::Plugin.register_input("kube_podinventory", self)

    @@hostName = (OMS::Common.get_hostname)`
    
  
The call to OMS::Common.get_hostname fails without it. Earlier this plugin was not the first plugin in the conf file and by the time it started running oms_common was already imported in global context. But since now its the first plugin we need to explicity import it

**Review by ganga1980 — APPROVED** [review]

**Review by jatakiajanvi12 — APPROVED** [review]
